barman/recovery_executor.py

```diff
@@ -1238,12 +1224,299 @@ def close(self):
+                    tablespace_src_path,
+                    tablespace_dst_path,
+                )
+                cmd_output = self.compression.uncompress(
+                    tablespace_src_path, tablespace_dst_path
+                )
+                _logger.debug(
+                    "Uncompression output for tablespace %s: %s",
+                    tablespace.name,
+                    cmd_output,
+                )
+        base_src_path = "%s/%s" % (staging_dir, base_file)
+        _logger.debug("Uncompressing base tarball from %s to %s.", base_src_path, dest)
+        cmd_output = self.compression.uncompress(
+            base_src_path, dest, exclude=["recovery.conf", "tablespace_map"]
+        )
+        _logger.debug("Uncompression output for base tarball: %s", cmd_output)
+
+    def _conf_files_exist(self, conf_files, backup_info, recovery_info):
+        """
+        Determine whether the conf files in the supplied list exist in the backup
+        represented by backup_info.
+
+        Returns a map of conf_file:exists.
+        """
+        exists = {}
+        for conf_file in conf_files:
+            source_path = os.path.join(recovery_info["destination_path"], conf_file)
+            exists[conf_file] = recovery_info["cmd"].exists(source_path)
+        return exists
+
+    def _copy_conf_files_to_tempdir(
+        self, backup_info, recovery_info, remote_command=None
     ):
+        """
+        Copy conf files from the backup location to a temporary directory so that
+        they can be checked and mangled.
+
+        Returns a list of the paths to the temporary conf files.
+        """
+        conf_file_paths = []
+
+        rsync = RsyncPgData(
+            path=self.server.path,
+            ssh=remote_command,
+            bwlimit=self.config.bandwidth_limit,
+            network_compression=self.config.network_compression,
+        )
+
+        rsync.from_file_list(
+            recovery_info["configuration_files"],
+            ":" + recovery_info["destination_path"],
+            recovery_info["tempdir"],
+        )
+
+        conf_file_paths.extend(
+            [
+                os.path.join(recovery_info["tempdir"], conf_file)
+                for conf_file in recovery_info["configuration_files"]
+            ]
+        )
+        return conf_file_paths
+
+
+def recovery_executor_factory(backup_manager, command, compression=None):
+    """
+    Method in charge of building adequate RecoveryExecutor depending on the context
+    :param: backup_manager
+    :param: command barman.fs.UnixLocalCommand
+    :return: RecoveryExecutor instance
+    """
+    if compression is None:
+        return RecoveryExecutor(backup_manager)
+    if compression == GZipCompression.name:
+        return TarballRecoveryExecutor(backup_manager, GZipCompression(command))
+
+    raise AttributeError("Unexpected compression format: %s" % compression)
+
+
+class Compression(with_metaclass(ABCMeta, object)):
```

Comment:
I keep wondering whether we could/should combine this with the `PgBaseBackupCompression` class introduced in an earlier commit. On one hand the hierarchy is similar and they both involve compression so it seems logical, but on the other hand they're dealing with different concerns so really they're different things.

Reply:
(I don't think this is a blocking change for this PR - just keeping the conversation regarding future refactorings alive).

Reply:
I think there might be something to do at larger scale. There are indeed those to classes (recovery_executor.Compression and compression.PgBaseBackupCompression) but there is also compression.Compressor
The main issue being in my opinion those class should have pure abstract class interfaces. It might help reveal common aspects while diverging ones should be handled without forcing heritage if it is not appropriate.